### PR TITLE
Python: Support for image content and chat message content with image content

### LINKED ISSUE #6536: Python: multi-modal input

### PR DESCRIPTION
**Pull request by eavanvalkenburg**
### Motivation and Context

Introduces a ImageContent class
Allows that type to be part of ChatMessageContent
Adds a sample for both local and online images
Adds many tests with data uri edge cases, and binary parsing edge cases.

Closes #6536 

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: python/semantic_kernel/contents/chat_message_content.py [hunk 1]
```diff
@@ -27,9 +29,10 @@
     TEXT_CONTENT_TAG: TextContent,
     FUNCTION_CALL_CONTENT_TAG: FunctionCallContent,
     FUNCTION_RESULT_CONTENT_TAG: FunctionResultContent,
+    IMAGE_CONTENT_TAG: ImageContent,
 }
 
-ITEM_TYPES = Union[TextContent, StreamingTextContent, FunctionResultContent, FunctionCallContent]
+ITEM_TYPES = Union[TextContent, StreamingTextContent, FunctionResultContent, FunctionCallContent, ImageContent]
```
Comment (moonbox3) [inline python/semantic_kernel/contents/chat_message_content.py:35]: Does this need a `Union` now? Can't we use the pipe `|`?

File: python/samples/concepts/chat_completion/image_input.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+from semantic_kernel import Kernel
+from semantic_kernel.connectors.ai.function_call_behavior import FunctionCallBehavior
+from semantic_kernel.connectors.ai.open_ai import AzureChatCompletion
+from semantic_kernel.contents import ChatHistory, ChatMessageContent, ImageContent, TextContent
+
+logging.basicConfig(level=logging.WARNING)
+
+system_message = """
+You are a image reviewing chat bot. Your name is Mosscap and you have one goal
```
Comment (moonbox3) [inline python/samples/concepts/chat_completion/image_input.py:14]: nit: `a image` -> `an image`

File: python/samples/concepts/chat_completion/image_input.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+critiquing images that are supplied.
+"""
+
+kernel = Kernel()
+
+service_id = "chat-gpt"
+chat_service = AzureChatCompletion(service_id=service_id)
```
Comment (moonbox3) [inline python/samples/concepts/chat_completion/image_input.py:21]: should we name this sample `azure_image_input` or something similar so we can have other examples with other models in the future?

File: python/semantic_kernel/contents/chat_message_content.py (hunk 1 above)
Comment (eavanvalkenburg) [reply python/semantic_kernel/contents/chat_message_content.py:35]: yes, this is a typevar and not a annotation!

File: python/semantic_kernel/contents/image_content.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    type: Literal[ChatMessageContentSubtypes.IMAGE_CONTENT] = Field(IMAGE_CONTENT_TAG, init=False)  # type: ignore
+    uri: Url | None = None
+    data: bytes | None = None
+    mime_type: str | None = None
+
+    @model_validator(mode="after")
+    def validate_uri_and_or_data(self) -> Self:
+        """Validate that either uri or data is provided."""
+        if not self.uri and not self.data:
+            raise ValueError("Either uri or data must be provided.")
```
Comment (moonbox3) [inline python/semantic_kernel/contents/image_content.py:60]: Shall we update this exception to be one of our SK defined exceptions?

File: python/tests/integration/completions/test_azure_oai_chat_service.py [hunk 5]
```diff
@@ -1,199 +1,183 @@
...
-
-    result: StreamingChatMessageContent | None = None
-    async for message in kernel.invoke_stream(function_name="chat", plugin_name="chat", arguments=arguments):
-        result = message[0] if not result else result + message[0]
-    output = str(result)
-
-    print(f"Math output: '{output}'")
-    assert "2" in output
-    assert 0 < len(output) < 500
+# @pytest.mark.asyncio
```
Comment (moonbox3) [inline python/tests/integration/completions/test_azure_oai_chat_service.py:4]: Is this file ready to be retired?

File: python/samples/concepts/auto_function_calling/chat_gpt_api_function_calling.py [hunk 6]
```diff
@@ -64,7 +64,7 @@
     temperature=0.7,
     top_p=0.8,
     function_call_behavior=FunctionCallBehavior.EnableFunctions(
-        auto_invoke=True, filters={"included_plugins": ["math", "time"]}
+        auto_invoke=False, filters={"included_plugins": ["math", "time"]}
```
Comment (juliomenendez) [inline python/samples/concepts/auto_function_calling/chat_gpt_api_function_calling.py:67]: Is this change on purpose? Asking because doesn't seem related to this PR, I thought it might be a local debugging change committed by mistake.

File: python/semantic_kernel/contents/image_content.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        if not self.uri and not self.data:
+            raise ContentInitializationError("Either uri or data must be provided.")
+        if self.uri and self.data:
+            logger.warning('Both "uri" and "data" are provided, "data" will be used.')
+        return self
+
+    def __str__(self) -> str:
+        """Return the string representation of the image."""
+        return (
+            f"data:{self.mime_type};base64,{ base64.b64encode(self.data).decode('utf-8')}"
```
Comment (juliomenendez) [inline python/semantic_kernel/contents/image_content.py:73]: nit: additional white space before `base64.b64encode`. It should be:
```python
f"data:{self.mime_type};base64,{base64.b64encode(self.data).decode('utf-8')}"
```

File: python/semantic_kernel/contents/image_content.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    @classmethod
+    def from_element(cls: type[_T], element: Element) -> _T:
+        """Create an instance from an Element."""
+        if element.tag != IMAGE_CONTENT_TAG:
+            raise ValueError(f"Element tag is not {IMAGE_CONTENT_TAG}")
+
+        if element.text:
+            return cls(
+                data=base64.b64decode(element.text.split(",")[1].encode()),
+                mime_type=element.text.split(",")[0].split(";")[0].split(":")[1],
```
Comment (juliomenendez) [inline python/semantic_kernel/contents/image_content.py:98]: Does all these splitting and reaching for specific indexes need to be guarded in `try/except`?

File: python/semantic_kernel/contents/streaming_text_content.py [hunk 9]
```diff
@@ -1,4 +1,10 @@
 # Copyright (c) Microsoft. All rights reserved.
+import sys
+
+if sys.version < "3.11":
+    pass
+else:
+    pass
```
Comment (juliomenendez) [inline python/semantic_kernel/contents/streaming_text_content.py:7]: :) what? :D :D

**Review by juliomenendez — CHANGES_REQUESTED** [review]

File: python/semantic_kernel/contents/image_content.py (hunk 8 above)
Comment (eavanvalkenburg) [reply python/semantic_kernel/contents/image_content.py:98]: I was thinking maybe replacing with a regex because this is indeed a bit clunky

File: python/samples/concepts/auto_function_calling/chat_gpt_api_function_calling.py (hunk 6 above)
Comment (eavanvalkenburg) [reply python/samples/concepts/auto_function_calling/chat_gpt_api_function_calling.py:67]: Good point, will revert

File: python/semantic_kernel/contents/streaming_text_content.py (hunk 9 above)
Comment (eavanvalkenburg) [reply python/semantic_kernel/contents/streaming_text_content.py:7]: Deleting the Self from the code, creates this, will remove

File: python/semantic_kernel/contents/image_content.py (hunk 7 above)
Comment (eavanvalkenburg) [reply python/semantic_kernel/contents/image_content.py:73]: I'll check, openai isn't picky about it 😏

**Review by moonbox3 — APPROVED** [review]

File: python/samples/concepts/chat_completion/azure_chat_image_input.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    async for message in answer:
+        print(str(message[0]), end="")
+    print("\n")
+
+
+async def main() -> None:
+    # print("Get a description of a image from a URL.")
+    # await chat(
+    #     uri="https://upload.wikimedia.org/wikipedia/commons/d/d5/Half-timbered_mansion%2C_Zirkel%2C_East_view.jpg"
+    # )
```
Comment (juliomenendez) [inline python/samples/concepts/chat_completion/azure_chat_image_input.py:72]: Do we need to keep this comment?

File: python/semantic_kernel/contents/utils/data_uri.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    @field_validator("parameters", mode="before")
+    def _validate_parameters(cls, value: list[str] | dict[str, str] | None = None) -> dict[str, str]:
+        if not value:
+            return {}
+        if isinstance(value, dict):
+            return value
+
+        new: dict[str, str] = {}
+        for item in value:
+            if item.strip() == "":
```
Comment (juliomenendez) [inline python/semantic_kernel/contents/utils/data_uri.py:103]: Would it make sense to do the follow up string manipulation on the stripped value? For example, could `item` be `"key1=val1   "`, then `item.split("=", maxsplit=1)` would return `["key1", "val1   "]`. Where if we store `item.strip()` in a variable and use it in line 103, 105, and 107, we get rid of that padding.

**Review by juliomenendez — APPROVED** [review]
Looks good, just 2 nitpicks.

**Review by moonbox3 — APPROVED** [review]

**Review by juliomenendez — APPROVED** [review]
